tests/math/test_is_independent.py

```diff
@@ -0,0 +1,499 @@
+            assert len(rnd_args) == num
+            torch.random.manual_seed(seed)
+            for _rnd_args in rnd_args:
+                expected = tuple(
+                    torch.rand(np.shape(arg)) * (bounds[1] - bounds[0]) + bounds[0] for arg in args
+                )
+                assert all(np.allclose(_exp, _rnd) for _exp, _rnd in zip(expected, _rnd_args))
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev, interface=interface)
+        def const_circuit(x, y):
+            qml.RX(0.1, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        constant_functions = [
+            const_circuit,
+            lambda x: np.arange(20).reshape((2, 5, 2)),
+            lambda x: (np.ones(3), -0.1),
+        ]
+
+        args_constant = [
+            (0.1, torch.tensor([-2.1, 0.1])),
+            (1.2,),
+            (torch.ones((2, 3)),),
+        ]
+
+        @qml.qnode(dev, interface=interface)
+        def nonconst_circuit(x, y, z):
+            qml.RX(0.1, wires=0)
+            qml.RY(y / 2, wires=0)
+            qml.RZ(qml.math.sin(z), wires=0)
+            return qml.expval(qml.PauliX(0))
+
+        nonconst_functions = [
+            nonconst_circuit,
+            torch.tensor,
+            lambda x: (1 + qml.math.tanh(1000 * x)) / 2,
+            *nonconst_lambdas,
+        ]
+
+        args_nonconst = [
+            (0.1, torch.tensor([-2.1, 0.1]), -0.9),
+            (-4.1,),
+            (torch.ones((3, 8)) * 1.1,),
+            *args_nonconst_lambdas,
+        ]
+
+        nonconst_expect_torch_fail = [False, False, False, *lambdas_expect_torch_fail]
+
+        @pytest.mark.parametrize("func, args", zip(constant_functions, args_constant))
+        def test_constant(self, func, args):
+            assert _is_independent(func, self.interface, args)
+
+        @pytest.mark.parametrize(
+            "func, args, exp_fail",
+            zip(nonconst_functions, args_nonconst, nonconst_expect_torch_fail),
+        )
+        def test_nonconst(self, func, args, exp_fail):
+            if exp_fail:
+                assert _is_independent(func, self.interface, args)
+            else:
+                assert not _is_independent(func, self.interface, args)
+
+        def test_kwargs_are_considered(self):
+            f = lambda x, kw=False: 0.1 * x if kw else 0.2
+            jac = lambda x, kw: torch.autograd.functional.jacobian(lambda x: f(x, kw), x)
+            args = (torch.tensor(0.2),)
+            assert _is_independent(f, self.interface, args)
+            assert not _is_independent(f, self.interface, args, {"kw": True})
+            assert _is_independent(jac, self.interface, args, {"kw": True})
+
+
+class TestOther:
+    """Other tests for _is_independent."""
+
+    def test_unknown_interface(self):
+        with pytest.raises(ValueError, match="Unknown interface: hello"):
+            _is_independent(lambda x: x, "hello", (0.1,))
```

Comment:
Very thorough testing!

I had a search though, and didn't find any `xfail` tests?

Reply:
Oh this is a very good comment, I messed something up there. That is, the `xfail` tests were removed at some point but then I found edge cases even for numerical+analytic testing. Will add them back in.